include/lsst/partition/ObjectIndex.h

```diff
@@ -0,0 +1,195 @@
+#define LSST_PARTITION_OBJECTINDEX_H
+
+// System headers
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+
+// LSST headers
+#include "lsst/partition/Csv.h"
+
+// Forward declarations
+namespace lsst {
+namespace partition {
+    class ChunkLocation;
+}} // namespace lsst::partition
+
+// This header declarations
+namespace lsst {
+namespace partition {
+
+/**
+ * Class ObjectIndex is a singleton serving as a front-end to a specific implementation of
+ * the "secondary" index.
+ * 
+ * Objects of the class can be open in two modes:
+ * - WRITE: for writing into an index file at the specified location.
+ * - READ: reading from an index via the specified source.
+ * 
+ * Files written in the WRITE mode are open in the "append" mode, hence their previoous
+ * content won't be truncated. It's up to a client to ensure the files get removed if
+ * the clean starting state is required.
+ * 
+ * The source specification in the READ node is a URL-style resource, which has the following
+ * general syntax:
+ * @code
+ *   <technology>:<technolog-specific-details>
+ * @code
+ * Presently, only the file-based technology is supported. Hence, the file URL is expected
+ * to look like this:
+ * @code
+ *   file:<path-to-the-index-file>
+ * @code
+ * Normally the path specified in the resource would be a file which was produced
+ * in the WRITE mode. The path could have a relative or absolute value. It could also be
+ * located on a local or remote filesystem. Since the file would be read sequentially then
+ * no specific requirements for file locking or direct access are imposed by this implemenation.
+ *
+ * Notes on the thread-safety of the class's implementation:
+ * 1) All public methods of the class are thread safe.
+ * 2) An internal state of the index is guarded by a mutex.
+ */
+class ObjectIndex {
+public:
+    // Objects of this class can't be copied or directly constructed by clients.
+    ObjectIndex(ObjectIndex const&) = delete;
+    ObjectIndex& operator=(ObjectIndex const&) = delete;
+
+    /// The non-trivial destructor is needed to close and release resources used by the index.
+    ~ObjectIndex();
+
+    /// Modes for opening the index.
+    enum Mode {
+        READ,
+        WRITE
+    };
+
+    /**
+     * The factory method.
+     * 
+     * @return  A reference to the underlining implementation of the index
+     */
+    static ObjectIndex& instance() {
+        static ObjectIndex oi;
+        return oi;
+    }
+
+    /// @return A value of 'true' if the index is open or created.
+    bool isOpen() const { return _isOpen; }
+
+    /// @return A mode the index was open/created in.
+    Mode mode() const { return _mode; }
+
+    /**
+     * Open (or create) index for writing into a local file.
+     * 
+     * @param fileName  A path to a local file to be open/created.
+     * @param editor  A CSV "editor" which is used for formatting output products.
```

Comment:
I don't want to do a premature over-design of this class since it's not quite clear in which direction this all will go in the future, or if the proposed partitioning mode would actually work on a large scale. Complicated design could be always added later.
Class `ObjecIndex` has two kinds of methods:
* implementation specific `open` and `create`
* and implementation-neutral methods `write` and `read`
At some point in the future, the first two may go into multi-class, factory, etc. implementation. And the last two will turn into pure virtual methods of some interface with multiple implementations behind it.

The `editor` is needed for extracting `objectId`, `chunkId` and `subChunkId` from the parsed input. It's used by the implementation of method `write`.

As I stated it earlier, the partitionier requires a serious re-factopring effort. We need to get rid of the complicated "map-reduce" machinery in favor of a more straight-forward "linear" parallelization logic. All those games with the static singleton were done excactly due to the current complexity of the "map-reduce".

Reply:
> Complicated design could be always added later.

Frankly I don't want complicated design later, I want a simple design now.

Reply:
The simple and streamlined design would have the following components:
* a `JSON`-based configuration class
* a class for reading (probably in many threads) input file(s) into in-memory buffers of a reasonably large size
* a class (much simple than `Silo`, etc. in the current "map-reduce" implementation) for writing output products into chunked files and overlaps
* an interface to the "secondary" index and its implementations for specific technologies (files, MySQL, Redis, etc.)
* the new  (simpler) `Worker` class which would be configured via its constructor with smart pointers to all above mentioned
* the application's `main` (or just some other function) which would instantiate all those mentioned above (including a desired number of workers), "introduce" instances to each other, and start processing in N threads.
Then there wouldn't be any need in statics, globals, etc.
In this design the partitioner could be easily turned into a library thingy to be embeddable in many copies within any host application, would the be some stand-alone tool, the Ingest system, or some permanent "Partitioning" service which would be pulling files from an Object Store over `http` and store its products back into the Object Store.
This would be the scalable architecture.